File: script/phys/ammo/World.ts
import { DebugDrawer } from "./DebugDrawer.js";
import { RayResultCallback } from "./RayResultCallback.js";
import { RigidBody } from "./RigidBody.js";
import { Vec3 } from "./Vec3.js";
import { impl as ammo } from "./context.js";

export class World {
    private _collisionConfiguration;
    private _broadphase;
    private _dispatcher;
    private _solver;

    readonly impl: any;

    private _debugDrawer: DebugDrawer | undefined = undefined;
    get debugDrawer(): DebugDrawer | undefined {
        return this._debugDrawer;
    }
    set debugDrawer(value: DebugDrawer | undefined) {
        this.impl.setDebugDrawer(value ? value.impl : ammo.NULL);
        this._debugDrawer = value;
    }

    constructor() {
        this._collisionConfiguration = new ammo.btDefaultCollisionConfiguration();
        this._dispatcher = new ammo.btCollisionDispatcher(this._collisionConfiguration);
        this._broadphase = new ammo.btDbvtBroadphase();
        this._solver = new ammo.btSequentialImpulseConstraintSolver();
        this.impl = new ammo.btDiscreteDynamicsWorld(this._dispatcher, this._broadphase, this._solver, this._collisionConfiguration);
    }

    addRigidBody(body: RigidBody) {
        this.impl.addRigidBody(body.impl)
    }

    rayTest(from: Vec3, to: Vec3, resultCallback: RayResultCallback): void {
        this.impl.rayTest(from.impl, to.impl, resultCallback.impl);
    }

    update(dt: number): void {
        this.impl.stepSimulation(1 / 60);
        const drawer = this.impl.getDebugDrawer();
        if (!ammo.compare(drawer, ammo.NULL)) {
            this.impl.debugDrawWorld();
        }
    }
}

File: puttyknife/phys/bullet-end/World.cpp
#include <phys/World.hpp>
#include <BulletCollision/CollisionDispatch/btDefaultCollisionConfiguration.h>
#include <BulletCollision/CollisionDispatch/btCollisionDispatcher.h>
#include <BulletCollision/BroadphaseCollision/btDbvtBroadphase.h>
#include <BulletDynamics/ConstraintSolver/btSequentialImpulseConstraintSolver.h>
#include <LinearMath/btIDebugDraw.h>

namespace phys
{
    class DebugDrawer : public btIDebugDraw
    {
    private:
        phys::DrawLineFunc _drawLineFunc;

    public:
        DebugDrawer(phys::DrawLineFunc &&drawLineFunc) : _drawLineFunc(std::move(drawLineFunc)) {}

        virtual void drawLine(const btVector3 &from, const btVector3 &to, const btVector3 &color) override
        {
            _drawLineFunc(from, to, color);
        }

        virtual void drawContactPoint(const btVector3 &PointOnB, const btVector3 &normalOnB, btScalar distance, int lifeTime, const btVector3 &color) override {}

        virtual void reportErrorWarning(const char *warningString) override {};

        virtual void draw3dText(const btVector3 &location, const char *textString) override {};

        virtual void setDebugMode(int debugMode) override {};

        virtual int getDebugMode() const override
        {
            return btIDebugDraw::DBG_MAX_DEBUG_DRAW_MODE;
        };

        ~DebugDrawer() {}
    };
}

phys::World *physWorld_new()
{
    btDefaultCollisionConfiguration *configuration = new btDefaultCollisionConfiguration();
    btCollisionDispatcher *dispatcher = new btCollisionDispatcher(configuration);
    btDbvtBroadphase *broadphase = new btDbvtBroadphase();
    btSequentialImpulseConstraintSolver *solver = new btSequentialImpulseConstraintSolver();
    return new btDiscreteDynamicsWorld(dispatcher, broadphase, solver, configuration);
}

int physWorld_stepSimulation(phys::World *world, float timeStep, int maxSubSteps, float fixedTimeStep)
{
    int res = world->stepSimulation(timeStep, maxSubSteps, fixedTimeStep);
    world->debugDrawWorld();
    return res;
}

void physWorld_addRigidBody(phys::World *world, phys::RigidBody *body)
{
    world->addRigidBody(body);
}

void physWorld_setDebugDrawer(phys::World *world, phys::DrawLineFunc &&drawLineFunc)
{
    btIDebugDraw *drawer = world->getDebugDrawer();
    if (drawer)
    {
        delete drawer;
    }

    drawer = new phys::DebugDrawer(std::move(drawLineFunc));
    world->setDebugDrawer(drawer);
}
